### ISSUE TITLE
mining: incorrect difficulty being reported

### ISSUE THREAD

**Issue by kilpatty** [handshake-org/hsd#150]
While building hstratum, we've run into some strange errors with the difficulty being reported as 0. 

Currently the Handshake difficulty is ~0.02083 which is calculated from a bits value of: 489684992. 

If we plug this into the difficulty function that exists in the mining common functions, we get a result of 0. 

https://github.com/handshake-org/hsd/blob/master/lib/mining/common.js#L82-L90

 It appears that the culprit is the Math.floor on line 89, which guarantees we will never get a decimal returned from this function. 

Is there a reason why that Math.floor exists here? 

I noticed there is an alternative difficulty function here: https://github.com/handshake-org/hsd/blob/62b4de584cdabd4488d0583cd4c2b2f3e75d4a60/lib/node/rpc.js#L2980-L2995 When we use this function, we are returned the correct difficulty. 

Should we swap out the difficulty function in the mining/common functions with that newer version? Will push up a PR shortly, but I wanted to see if there was a reason for the Math.floor that currently exists there.

**Comment by kilpatty** [comment]
Okay so I went a bit down the rabbit hole on this one....

I wrote a quick gist to show the variance in difficulties based on the two separate functions: https://gist.github.com/kilpatty/0148cc09c1f7cc96454147a1589cfd28

From the above gist we can see that the Math.floor function is causing the difficulty to be rounded down which gives us slightly different numbers in difficulty. Although, due to Handshake having such a low network difficulty at the moment we are rounding down to 0 which seems to be a problem. 

I also investigated how Bitcoin Core is running this calculation, and it looks like they are using the same function as the rpc is using (toDifficulty): https://github.com/bitcoin/bitcoin/blob/3515612e069e3730f173a08e60b99f96b174188d/src/rpc/blockchain.cpp#L60-L80

I'm looking for some input on how we should solve this: 

A. We change the common.getDifficulty function to use the one specified in the rpc file.
B. We remove the Math.floor from the common.getDifficulty file
C. We remove difficulty entirely from block template (Currently the only place the mining/common.getDifficulty is used), and the mining/common.getDifficulty function entirely. 

Per C -> We expose the difficulty in the block template, but it is not used in very many places mainly just the following: https://github.com/handshake-org/hsd/blob/af86d4899791601d10212115c80a2f2136fdb004/lib/mining/miner.js#L166 (log statement)

@tynes @pinheadmz Curious on your guys' thoughts on this, and how we should tackle it. Happy to push up a PR to fix this, but just interested in which direction you think is better. This also affects bcoin as well I believe.

**Comment by pinheadmz** [comment]
First reaction: https://bitcoin.stackexchange.com/questions/50857/is-the-difficulty-target-always-truncated-for-practical-purposes

...I'm gonna go down this rabbit hole too, but also note in bcoin the `common.getDifficulty()` method also uses `Math.floor`

**Comment by kilpatty** [comment]
I think that stack exchange issue is talking about target truncation rather than difficulty rounding if I'm reading it correctly. https://bitcoin.stackexchange.com/questions/76560/why-is-the-target-truncated

I actually stumbled across this in bcoin first as I am forking bstratum to hstratum, and we were running into a lot of issues of the difficulty being set by the pool being 0 (due to the rounding down issue when difficulty was really .00283). 

My gut instinct is that because getDifficulty isn't actually storing anything in "consensus", it's just a helper function, that it should be as precise as possible. But, I'm not sure if there is something that I'm missing entirely that is justifying the Math.floor option. 

At the very least is seems that we should add an if statement where Math.floor is not called if difficulty < 1, so that it's never reported as 0.

**Comment by pinheadmz** [comment]
The actual PoW in hsd gets verified here:
https://github.com/handshake-org/hsd/blob/master/lib/protocol/consensus.js#L461-L483

...which replicates the Bitcoin Core method here:
https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp#L74-L91

And the PoW target is calculated here:
hsd:
https://github.com/handshake-org/hsd/blob/master/lib/blockchain/chain.js#L2863-L2907

Bitcoin Core:
https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp#L13-L47

**Comment by pinheadmz** [comment]
*that last Bitcoin Core link is the source of `nBits` when Core mines a block as well:
https://github.com/bitcoin/bitcoin/blob/master/src/miner.cpp#L166

**Comment by kilpatty** [comment]
Just to clarify the getDifficulty function has no impact on consensus/PoW code at all. It's just a helper function exposed in the miner package.

**Comment by pinheadmz** [comment]
Yeah, I think I'm following you, I'm not sure what `mining/common.getDifficulty()` is really for yet. But I don't think its used for *actual* mining ...?

**Comment by kilpatty** [comment]
No not used for actual mining, but I think it's slightly deceptive in that if you want to use bcoin/hsd to build a block template and use that in external libraries you could get slight errors from relying on this difficulty measurement. 

e.g. https://github.com/bcoin-org/bstratum/blob/51a9aa0b4dbe2a0c963835bf22221a96e00ec01d/lib/stratum.js#L571-L573 bstratum relies on getDifficulty to get the validity of a share. If a client submits a share with difficulty of 1.5, it may be rounded down and assumed to be too low, and therefore invalidate that specific share.

**Comment by pinheadmz** [comment]
@kilpatty Is this still an issue? almost a year out now.

**Comment by pinheadmz** [comment]
@kilpatty ping. Is this still an issue? Or was there just confusion from the API?

**Comment by pinheadmz** [comment]
Ok, so I think this issue can be closed. It's true that `Marh.floor()` seems out of place there - but it is ONLY used in the logging message anyway. It's also true that if the difficulty is low enough, you'll end up with a `0` there. But now we're a year into mainnet, and we can start using big-boy (and -girl) grown-up numbers.


I did a little test to compare `mining/common.getDifficulty(target)` with `node/rpc.toDifficulty(bits)` like so:

1. At chain height `52336` I call `rpc getblocktemplate`:

```
Created block tmpl (height=52337, weight=75271, fees=1.7883, txs=77, diff=827604230, bits=419770503).
```

So, difficulty is represented as a rounded-down int `827604230` in the message, representing what the **next** block's difficulty target will be (remember we adjust every block) as computed by `mining/common.getDifficulty(target)`

2. Wait... like half an hour... for a real-life mainnet miner to actually find block `52337` and then call `hsd-rpc getdifficulty`:

```
827604230.9046377
```

This is the difficulty calculated by `node/rpc.toDifficulty(bits)` based on the current chain tip (block `52337`)

So they are equal. And in fact, before I ran this test I even modified `mining/common.getDifficulty(target)` to `console.log` the NON-rounded result and it was identical to the eventual result from the rpc method after the real block was received: `827604230.9046377`

So, @kilpatty I'm gonna put this one-and-a-half-year-old issue to rest. If you wanna remove the `Math.floor()` for the log message I understand and a PR is okay for that, but I think as far as the title of this issue is concerned, we can close.
